src/collapse/mod.rs

```diff
@@ -69,3 +85,183 @@ pub trait Collapse {
     /// - `Some(false)` means "no, this implementation definitely won't work"
     fn is_applicable(&mut self, input: &str) -> Option<bool>;
 }
+
+///////////////////////////////////////////////////////////////////////
+
+/// Occurrences is a HashMap, which uses:
+/// * Fnv if single-threaded
+/// * CHashMap if multi-threaded
+#[derive(Clone)]
+enum Occurrences {
+    SingleThreaded(FnvHashMap<String, usize>),
+    MultiThreaded(Arc<CHashMap<String, usize>>),
+}
+
+impl Occurrences {
+    fn new(nthreads: usize) -> Self {
+        if nthreads <= 1 {
+            let map = FnvHashMap::with_capacity_and_hasher(
+                CAPACITY_SINGLE_THREADED_HASHMAP,
+                fnv::FnvBuildHasher::default(),
+            );
+            Occurrences::SingleThreaded(map)
+        } else {
+            let map = CHashMap::with_capacity(CAPACITY_MULTI_THREADED_HASHMAP);
+            let arc = Arc::new(map);
+            Occurrences::MultiThreaded(arc)
+        }
+    }
+
+    fn add(&mut self, key: String, count: usize) {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(map) => *map.entry(key).or_insert(0) += count,
+            MultiThreaded(arc) => {
+                arc.upsert(key, || count, |v| *v += count);
+            }
+        }
+    }
+
+    fn is_concurrent(&self) -> bool {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(_) => false,
+            MultiThreaded(_) => true,
+        }
+    }
+
+    fn write<W>(&mut self, mut writer: W) -> io::Result<()>
```

Comment:
Should this just consume `self` instead? That way you can use `Arc::try_unwrap` below, and won't need the `mem::replace`.

Reply:
After we write out the contents of the map, we want the `Folder` to continue to stay alive. Said differently, `collapse` take a `&mut self`; so either we do `mem::replace` here or we do it in the collapse function. I'd just assume keep it here instead of transferring similar logic to the `collapse` method.

Reply:
Ah, in that case I'd prefer it be called `write_and_clear` to make that obvious.

Reply:
Alternatively, could we have it _not_ clear and just operate on borrowed values instead?

Reply:
Yah - Renaming to `write_and_clear` is better. Unfortunately, though, operating on borrowed values is not possible. That's what I had initially hoped to do, but the concurrent hashmap doesn't give you an iterator over borrowed values. I'm assuming it has something to do with how it's implemented that prevents that from being safe. For the concurrent hashmap, we have to drain it to get access to the data. And since we should do the same thing regardless of whether we've got a single-threaded or multi-threaded hashmap, I think we should stick with the current method, just renamed to better reflect what's going on.